docs(0028): update spec with monetization tag observation changes

sublimator opened the pull request:
Overview
* Specify observing head for monetization tags
  * Introduce concept of a meta "Tags Set"
    * if tags are injected they must be ordered carefully to work with meta observer
* Rename Request-ID to Monetization-ID
  * leave `requestId` field the same for the moment
    * Add comment near field for correlation
* Consistently use `browser` rather than mix of `browser` and `agent`
* Bar iframe support until have a sensible working implementation
* meta[content] can specify a payment pointer or an SPSP url
  * Generically referred to as `Payment Setup Endpoint`
* Make it clear that the amount is the destination received amount

sublimator commented:
@sharafian

sublimator commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -44,24 +44,30 @@ This flow refers to the user's **browser** and the user's **provider**, [defined
+  - The event's type is `monetizationstart`. The event has a `detail` field with an object containing the Payment Pointer and the Monetization ID ([specified below](#monetizationstart)).
   - The user's agent also emits a `monetizationprogress` ([specified below](#monetizationprogress)) event from `document.monetization`, corresponding to this first packet. If there are no listeners the event MAY NOT be emitted.
-- Payment continues until the user closes/leaves the page.
+- Payment continues for the lifetime of a given meta tag (or semantically equivalent)
```
should really be a set of meta tags I guess, but yeah ...

sharafian commented:
We should also have some tips for implementors about CORS and the Web-Monetization-Id header. This is something I didn't even think about at first because extensions ignore CORS but in some cases you need to serve an OPTIONS endpoint that has Access-Control-Allow-Headers set to allow Web-Monetization-Id

sharafian commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -140,12 +136,12 @@ Dispatched once the first ILP packet with a non-zero amount has been fulfilled b
   detail: {
     paymentPointer: String,
-    requestId: String
+    id: String
```
What's the rationale for changing this field? This turns this into a breaking change

sublimator commented:
Putting aside the fact it is a breaking change:
I think "id" and  "Monetization ID" are more appropriate. The former is
nice and short (leaving wiggle room?? Actually, I could be easily moved to
s/requestId/monetizationId) and the latter matches the existing
Web-Monetization-ID header.

The meta tags and are no longer bound to a single page "request" and
streams can stop and start.

Back to "breaking":
Could we live with the pain of the change? Is it worth it for conceptual
clarity? How many people would it effect? When would be the time to make
changes like this?

Does "request" still make sense to you? Would it make sgnse to the
uninitiated?

I guess one could define non enumerable properties on the events with
deprecation warnings. Shrug.

Your call

sharafian commented:
> The meta tags and are no longer bound to a single page "request" and
streams can stop and start.

The main reason to say requestId was to make it clear that the ID was ephemeral (i.e. unique to each SPSP request). A monetizationId might seem like it refers to the user rather than to the SPSP request.

I could maybe get `webMonetizationId` (because it matches the header) or just `id` because it's pretty general. But I still feel like `requestId` isn't innapropriate.

> - monetization stop event ?

:+1: I think this makes sense. We might also want to add a new monetization state: `stopped`. So you can tell the difference between initializing (`"pending"`), stopped (`"stopped"`), and not supported (`undefined`). This also makes it less confusing when you check if web monetization is supported on a page that doesn't use it. "pending" is a confusing state if there's no meta tag present.

> - specify both sent and received currency code/scale explicitly

:+1: STREAM exposes both of these so it makes sense to expose them to the page. You'd want to use the source currency preferences for display (like counters) and you'd want to use the destination currency preferences to check how much the user has paid (e.g. to adjust paid video quality or to trigger loading of a piece of exclusive content)

sublimator commented:
@sharafian 
> just id because it's pretty general.

Clarity through ambiguity TM :)

@traviscrist 
> it no longer feels short lived,

I think I understand. 

We moved to allow dynamic `<meta>` tags as a way of supporting SPA, such as coil.com (which I suppose will end up using server side rendering for SEO, but regardless, I'm sure you can imagine the use cases, such as a long page of videos that stop/start as you scroll ).  

----

For better or worse, I suppose I've come to think of a `monetization` as a concept/thing.

I'll try and illustrate why `monetizationId` made sense to me:

The event names:
**monetization**start 
**monetization**stop

The meta tag:
`<meta name='`**`monetization`**`' />`

The (content request) header:
Web-**Monetization-Id**

Look at the relationships between the various entities:

|Entities|Relationship|
|--------|------------|
|`<meta name="monetization">`:requestId| 1:1|
|SPSP resolve request: requestId | M:1 (see: pause/resume)| 
|Content request: requestId | M:1 (see: reconnects while streaming videos) |

When do we generate a new `requestId` now that we have moved to supporting dynamic meta tags? Whenever we see a new monetization meta tag.

Note, that in the table above I'm using request in the sense of a http request, rather than in a more abstract sense. Kind of cheating, but there *is* a strict mapping of meta tag to an identifier. To me the current `request` in `requestId` maps more closely to the request for content than the SPSP request.

I recall that requestId was at first dubbed `correlationId`. Used for correlating a monetization stream with later requests for content/services/etc. 

If you guys aren't really board, given it's a break change, should probably just leave it as `requestId`

sublimator commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -44,24 +44,30 @@ This flow refers to the user's **browser** and the user's **provider**, [defined
+  (TODO: This would imply that when dynamically injecting multiple meta tags (in the sense of others than `<meta name="monetization">`, it would require careful ordering so as to be operating on the information defined by the set of tags as a whole, removal with document.head.querySelectorAll('meta[name^="monetization"]'), then adding the name="monetization" tag last)
+  
+  )
+  - The browser will generate a fresh UUID (version 4) and use this as the Monetization ID from this point forward. 
```
@traviscrist 
Yes, this needs fixing ^

sublimator commented:
@sharafian @traviscrist 
Thoughts re: monetizationpause/ monetizationresume events?
Would be interesting to get input from cinnamon guys.

Note that often people open a youtube tab to put some music on, change tabs while still listening and then do something else.

sublimator commented:
@sharafian 
> We should also have some tips for implementors about CORS
> "pending" is a confusing state if there's no meta tag present.

Noted

sublimator commented:
@traviscrist 
Do you have any specific concerns with easing the requirement of only one meta per "page load" (which I assume means a single http request)? In particular, any security concerns? That was one aspect that nags at me.

We were discussing an https requirement at one point, but then you'd really have to be the browser to be able to get to the raw request bytes to ensure it wasn't inserted dynamically. Failing that, you'd have to request the page again, which I doubt would be workable in all cases.

sublimator commented:
YouTube question wasn't really concerning the correlation id. Just
something that occurred to me. The extension currently pauses monetization
when a tab is not active.

We moved the random id generation from per page load to per meta content
change.

Late here, will answer tomorrow in more detail

sublimator commented:
> So this means that if the page is reloaded then the id stays the same?
No. The requirements were not intended to be relaxed. It's just that the
language needs tightening up.

The idea (and latest impl in the extension) is to generate a new uuid.v4
every time it sees a new meta tag, and it's not persisted between page
loads. Reload the page and you'll get a new id.

sublimator commented:
> I don't think that's very clear in the updated documentation
Agreed. Still very rough. Thanks for highlighting it

> I'm also good with calling it monetizationId
@sharafian 
What about you?

sublimator commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -44,24 +44,30 @@ This flow refers to the user's **browser** and the user's **provider**, [defined
+  - The `<meta>` tags MUST be in the top level window (i.e. not inside an iframe)  
+  - If any of the Web Monetization `<meta>` tags are malformed, the browser will stop here. The browser SHOULD report a warning via the console.
+  - If the Web Monetization `<meta name="monetization">` tag is well-formed, the browser should extract the Payment Pointer or SPSP url
+  (TODO: This would imply that when dynamically injecting multiple meta tags (in the sense of others than `<meta name="monetization">`, it would require careful ordering so as to be operating on the information defined by the set of tags as a whole, removal with document.head.querySelectorAll('meta[name^="monetization"]'), then adding the name="monetization" tag last)
```
That's just some scribble thoughts I put in for discussion/thought ... 

People will have to be careful to make sure they have all their ducks (monetization meta tags used to convey info other than a payment pointer) in a row before updating.

sublimator commented:
@sharafian 
> Payment Pointer or SPSP url

Do we have a single term that covers both ?
Note that the emitted monetization events currently use `paymentPointer` 
If not, do we need to constantly put `Payment Pointer or SPSP url` or just a single well placed mention that it's possible to also use an SPSP url. 

Does the fact that you might need to use an SPSP url mean the design of payment pointers is lacking somehow?

sublimator commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -41,33 +41,37 @@ Web Monetization makes use of [SPSP](../0009-simple-payment-setup-protocol/0009-
-- The browser sets `document.monetization` to an Object which implements [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget).
-- The browser sets `document.monetization.state` to `pending`.
-- The browser observes the `<head>` and looks for a Web Monetization `<meta name="monetization">` tag ([specified below](#meta-tags)).
+- The user's agent navigates a webpage.
```
I changed these to consistently use `user's agent` though perhaps it should have been the other way around?

sharafian commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -41,33 +41,37 @@ Web Monetization makes use of [SPSP](../0009-simple-payment-setup-protocol/0009-
-  - Each time a packet with a nonzero amount is fulfilled, the provider notifies the browser, and the browser emits an event on `document.monetization`. The event's type is `monetizationprogress`. The event has a `detail` field containing the details of the packet ([specified below](#monetizationprogress)). If there are no listeners the event MAY NOT be emitted.
-  - When a stream is closed the `document.monetization.state` MUST be set back to 'pending'
+
+  - The user's agent generate a fresh **random** UUID (version 4) and uses this as a Monetization ID. This MUST be regenerated upon dynamic changes to the meta tags and MUST be unique per page load in order to avoid user tracking.
```
it's fine to call it out explicitly. There's a grammatical error though: `generate a fresh random UUID` should be `generates a fresh random UUID`

sublimator commented:
@sharafian
I believe that was already fixed in a subsequent fix

sharafian commented:
Is this PR blocked on anything? @sublimator @traviscrist

sublimator commented:
I think the plan was to get experience with things working in the extension
and then make this PR match that.

sharafian commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -139,13 +139,15 @@ Dispatched once the first ILP packet with a non-zero amount has been fulfilled b
   detail: {
+    // Payment Setup Endpoint
     paymentPointer: String,
+    // Monetization ID
```
I'm against renaming fields unless we absolutely have to

sublimator replied:
Also added Web-Monetization-ID header to the comment

sharafian commented on 0028-web-monetization/0028-web-monetization.md:
````diff
@@ -154,13 +156,18 @@ Dispatched every time an ILP packet with a non-zero amount has been fulfilled by
 }
 ```
 
+(TODO: 
````
Let's make it clear that these are the receiving amounts then. We could introduce sent amounts addititively in a later update but for now I think we just need to get the spec up to parity with the implementation

sublimator commented:
This is still very much a draft, hence the TODOs. I guess we could just
merge this in now after a bit of cleanup though as it more closely matches
reality and it is taking time. I didn’t understand Ben’s question about
“what’s the hold up?” properly before.

As for a follow up:
I think there are ways we could change the names, while keeping the old
fields (non enumerable) which console.warn deprecation messages upon access
if in our heart of hearts we want to change some names.I think the
amount/asset etc, could do with clarifying specification of src/dest. I’d
do it now/soon before much adoption.

sublimator commented:
I watched the Sublime Text 2 to Sublime Text 3 transition, which went from
Python 2 to Python 3 and a naming convention change for the plugin api.
There was a transition period with compatibility layers but it was possible
to make a clean break.

Changes from v1 to v2 included using json for conf over xml. There were far
less users then.

In the end it was worth it.

I don’t think we should be afraid to make changes in the early days,
especially if it can be done in a non breaking way.

As for ‘requestId’, it’s kind of grown on me/could live with it, but that’s
typical once you no longer have fresh eyes.

adrianhopebailie commented:
@sublimator @sharafian @traviscrist 

This spec is likely to move from here to something in the W3C format as soon as @brucelawson and I have submitted this to the Web Platform Incubator CG (WICG) later this week.

I'd recommend we look at already migrating this over to that doc (incorporating these changes), linking to it from here, deprecating this and making further changes there.

@sublimator - do you want to have a stab at that or would you prefer I do it?

The WICG has a template for new proposals which I've already used to put the following together: https://github.com/adrianhopebailie/web-monetization

The main entry point for the community is the explainer which is expected to be updated regularly based on feedback and then the spec (index.html) should be updated whenever there is consensus on changes.

The spec will need to be a balance of "what is implemented already" and "what the community says should be in the spec". i.e. Initially the implementation drives the spec but as we get more implementations the spec will become the reference and implementations will need to follow.

Keep in mind that the spec is targeted at browser vendors (they are the implementors) not web developers. BUT, the API design should make sense to developers. So, for example, the choice of property names should consider only what the developer cares about (i.e. not SPSP, STREAM or ILP).

sublimator commented:
@adrianhopebailie

I defer to your greater expertise here.
I'm sure Ben and Travis will want to have some input before it's 'thrown to
the wolves' though

sharafian commented:
@adrianhopebailie Right now there's no accurate spec of WM other than this PR so I think it's important to get this PR merged. Your spec that you linked looks good, I'm wondering where I can leave feedback on it

sublimator commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -41,33 +41,39 @@ Web Monetization makes use of [SPSP](../0009-simple-payment-setup-protocol/0009-
 
 This flow refers to the user's **browser** and the user's **provider**, [defined above](#terminology).
```
*browser* vs *agent*

sublimator commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -41,33 +41,39 @@ Web Monetization makes use of [SPSP](../0009-simple-payment-setup-protocol/0009-
+    - On the SPSP query to resolve the Payment Setup Endpoint, a `Web-Monetization-Id` header is sent, containing the Monetization ID. The server running the web-monetized site may use this to associate future requests by the user with their payments.
+    - With the `destination_account` and `shared_secret` fetched from the SPSP query, a STREAM connection is established. A single STREAM is opened on this connection, and a positive SendMax is set.
+    - The provider SHOULD set their SendMax high enough that it is never reached, and make payment adjustments by limiting throughput.
+  - Once the STREAM connection has fulfilled an ILP packet with a non-zero amount, the provider notifies the browser:
```
@sharafian @traviscrist 

The phrase:  "the provider notified (grammar error?) the browser, and the browser dispatches an event on `document.monetization`."  kind of threw me for bit, as I misread it as "provider notifies browser VIA dispatching an event on document.monetization"

In terms of "the provider notified the browser":

The oauth polyfill just "does it" and the "provider" in the case of the extension sends a wext runtime 'monetizationProgress' message to the content script.

adrianhopebailie replied:
This is the as-yet-undefined interface from provider to browser that we need to  design and implement more explicitly.

I don't think we need to do that in this PR but we should start thinking about soon and include in the W3C proposal.

sublimator replied:
Take a crack :) 
I might have a go tomorrow or the weekend. 
Sleepy McJoe today :)

sublimator commented on 0028-web-monetization/0028-web-monetization.md:
```diff
@@ -139,13 +134,15 @@ Dispatched once the first ILP packet with a non-zero amount has been fulfilled b
-    requestId: String
+    // Payment Setup Endpoint
+    paymentPointer: string,
+    // Monetization ID / Web-Monetization-ID header
```
Added these comments here for correlation

adrianhopebailie commented:
@sharafian 

> Right now there's no accurate spec of WM other than this PR so I think it's important to get this PR merged.

Agree, let's get it merged and then we can migrate it to the W3C template. Can we aim to have that done this week?

> Your spec that you linked looks good, I'm wondering where I can leave feedback on it

The WICG process is to have a dedicated repo for proposals so use the issue list on that repo to log any comments about the spec or the explainer.

sharafian commented:
Merging on behalf of Niq
